spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -549,76 +390,185 @@ public Object doWithRetry(RetryContext context) {
+									}
+									else {
+										if (KafkaMessageChannelBinder.this.logger.isDebugEnabled()) {
+											KafkaMessageChannelBinder.this.logger.debug(
+													"Sent to DLQ " + messageLog.toString());
+										}
+									}
 								}
-							}
-						}
-					});
+							});
 				}
 			});
 		}
-		kafkaMessageDrivenChannelAdapter.start();
+		return kafkaMessageDrivenChannelAdapter;
+	}
 
-		EventDrivenConsumer edc = new EventDrivenConsumer(bridge, rh) {
+	@Override
+	protected MessageHandler createProducerMessageHandler(final String destination,
+			ExtendedProducerProperties<KafkaProducerProperties> producerProperties) throws Exception {
+		ProducerMetadata<byte[], byte[]> producerMetadata = new ProducerMetadata<>(destination, byte[].class,
+				byte[].class,
+				BYTE_ARRAY_SERIALIZER, BYTE_ARRAY_SERIALIZER);
+		producerMetadata.setSync(producerProperties.getExtension().isSync());
+		producerMetadata.setCompressionType(producerProperties.getExtension().getCompressionType());
+		producerMetadata.setBatchBytes(producerProperties.getExtension().getBufferSize());
+		Properties additional = new Properties();
+		additional.put(ProducerConfig.ACKS_CONFIG, String.valueOf(this.configurationProperties.getRequiredAcks()));
+		additional.put(ProducerConfig.LINGER_MS_CONFIG,
+				String.valueOf(producerProperties.getExtension().getBatchTimeout()));
+		ProducerFactoryBean<byte[], byte[]> producerFB = new ProducerFactoryBean<>(producerMetadata,
+				this.configurationProperties.getKafkaConnectionString(), additional);
+		final ProducerConfiguration<byte[], byte[]> producerConfiguration = new ProducerConfiguration<>(
+				producerMetadata, producerFB.getObject());
+		producerConfiguration.setProducerListener(this.producerListener);
+		KafkaProducerContext kafkaProducerContext = new KafkaProducerContext();
+		kafkaProducerContext.setProducerConfigurations(
+				Collections.<String, ProducerConfiguration<?, ?>>singletonMap(destination, producerConfiguration));
+		return new ProducerConfigurationMessageHandler(producerConfiguration, destination);
+	}
 
-			@Override
-			protected void doStop() {
-				// stop the offset manager and the channel adapter before unbinding
-				// this means that the upstream channel adapter has a chance to stop
-				kafkaMessageDrivenChannelAdapter.stop();
-				if (messageListenerContainer.getOffsetManager() instanceof DisposableBean) {
-					try {
-						((DisposableBean) messageListenerContainer.getOffsetManager()).destroy();
+	@Override
+	protected void createProducerDestinationIfNecessary(String name,
+			ExtendedProducerProperties<KafkaProducerProperties> properties) {
+		if (this.logger.isInfoEnabled()) {
+			this.logger.info("Using kafka topic for outbound: " + name);
+		}
+		validateTopicName(name);
+		Collection<Partition> partitions = ensureTopicCreated(name, properties.getPartitionCount());
+		if (properties.getPartitionCount() < partitions.size()) {
```

Comment:
this logs a message, but in the line above the smaller count was passed to `ensureTopicCreated`... I admit I'm not fully familiar with the flow (yet), but wanted to point this out in case it's incorrect... not real clear when reading this code in any case - maybe the log message should be closer to where the other value supercedes